### PR TITLE
[Dubbo-5961] Fix disorder problem of notification

### PR DESCRIPTION
**Pull request by laddcn**
## What is the purpose of the change

Fix the bug of notification order mentioned in https://github.com/apache/dubbo/issues/5961 

## Brief changelog

1. Fix disorder problem of notification
2. Add version control of notification data  and avoid useless notify by version comparation
3. Fix bugs with FailedNotifiedTask
4. Add test cases for retry notify

## Verifying this change

Test cases run with success for for retry notify

Follow this checklist to help us incorporate your contribution quickly and easily:

- [x] Make sure there is a [GITHUB_issue](https://github.com/apache/dubbo/issues) field for the change (usually before you start working on it). Trivial changes like typos do not require a GITHUB issue. Your pull request should address just this issue, without pulling in other changes - one PR resolves one issue.
- [x] Format the pull request title like `[Dubbo-XXX] Fix UnknownException when host config not exist #XXX`. Each commit in the pull request should have a meaningful subject line and body.
- [x] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
- [x] Write necessary unit-test to verify your logic correction, more mock a little better when cross module dependency exist. If the new feature or significant change is committed, please remember to add sample in [dubbo samples](https://github.com/apache/dubbo-samples) project.
- [x] Run `mvn clean install -DskipTests=false` & `mvn clean test-compile failsafe:integration-test` to make sure unit-test and integration-test pass.
- [ ] If this contribution is large, please follow the [Software Donation Guide](https://github.com/apache/dubbo/wiki/Software-donation-guide).

### CODE REVIEW

**Comment by wavesZh** [comment]
If using version  comparison, where version should be apply from, consumer or provider? If consumer perceive provider's changes by pull mode, eg Zookeeper or others, with network latency. Maybe i'm sensitive

**Comment by laddcn** [comment]
Yes, you are right. The best way is to tag version when data are produced in provider side(or dubbo admin which will produce config info). However, it is difficult to to implement:
1.	Since providers and dubbo admin may have many instances, we need import a distributed id generator (such as leaf https://tech.meituan.com/2017/04/21/mt-leaf.html) to generate version id. 
2.	Dubbo version compatibility is also very difficult. Application with lower dubbo version will not produce data with version id. And it is almost impossible to update all applications in company.
3.	Codes with data production distribution in various projects. And it is also difficult to modify these codes to tag data version.

Notification data change no often. And it occurs rarer in a very short time which may lead disorder problem by network latency. So the problem you mentioned above occur rarely.

In my opinion, it is not worth to tag version in provider side. I have produce tag in consumer side now.

**Comment by chickenlj** [comment]
Does FailedNotifiedTask need to keep the address list it used when fail? What if we change FailedNotifiedTask to only stands for one event and notify the latest address list.

**Comment by chickenlj** [comment]
Now, I doubt the design of the notification retry. In what case does a notification need to be retried, every exception in every step is predictable and they will happen again if they happened for the first time.

File: dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/retry/FailedNotifiedTask.java [hunk 1]
```diff
@@ -58,10 +63,31 @@ public void removeRetryUrl(List<URL> urls) {
...
+            try {
+                registry.doNotify(url, listener, urls);
+
+                //When notify with success, clear urls and stop this task
+                urls.clear();
+                this.cancel();
+                logger.info("Retry to notify with success for subscribe " + url);
+            } catch (Exception e) {
+                //When notify with failure, retry again
+                reput(timeout, retryPeriod);
```
Comment (chickenlj) [inline dubbo-registry/dubbo-registry-api/src/main/java/org/apache/dubbo/registry/retry/FailedNotifiedTask.java:88]: There's reput logic in the parent class.

**Comment by chickenlj** [comment]
Closing

**Pull request closed without merging**